thorlcr/thorutil/thbuf.cpp

```diff
@@ -606,6 +608,549 @@ class CSmartRowInMemoryBuffer: public CSimpleInterface, implements ISmartRowBuff
+    {
+#ifdef VERIFY_ROW_IDS_SPILLING_ROWSTREAM
+        unsigned id;
+        memcpy(&id, row, sizeof(unsigned));
+        assertex(id-1 == expectedId);
+        trace("READ: %s nextInputRow: %" RCPF "u", msg, expectedId);
+#endif
+    }
+    const void *readToMarker()
+    {
+        rowcount_t markerRow = std::get<1>(readFromStreamMarker);
+        if (markerRow == nextInputRow)
+        {
+            const void *ret = std::get<0>(readFromStreamMarker);
+            checkCurrentRow("M: ", ret, nextInputRow);
+            readFromStreamMarker = { nullptr, 0, 0 };
+            readState = rs_fromqueue;
+            ++nextInputRow;
+            return ret;
+        }
+        else if (markerRow > committedRows) // rows we need have not yet been committed to disk
+        {
+            CLeavableCriticalBlock b(readerWriterCS);
+            if (markerRow > committedRows) // JCSMORE this is waiting too long, I don't need markerRow to be committed, I need nextInputRow to be committed ..
+            {
+                // wait for writer to commit
+                readerWaitingForCommit = true;
+                b.leave();
+                trace("READ: waiting for committedRows(currently = %" RCPF "u) to catch up to markerRow(%" RCPF "u), nextInputRow = %" RCPF "u", committedRows.load(), markerRow, nextInputRow);
+                moreRows.wait();
+                assertex(markerRow <= committedRows);
+            }
+        }
+        const void *row = readRowFromStream();
+        ++nextInputRow;
+        return row;
+    }
+public:
+    IMPLEMENT_IINTERFACE_O_USING(CSimpleInterfaceOf<ISmartRowBuffer>);
+
+    explicit CCompressedSpillingRowStream(CActivityBase *_activity, const char *_baseTmpFilename, bool _grouped, IThorRowInterfaces *rowIf, const LookAheadOptions &_options, ICompressHandler *_compressHandler)
+        : activity(*_activity), baseTmpFilename(_baseTmpFilename), grouped(_grouped), options(_options), compressHandler(_compressHandler),
+          meta(rowIf->queryRowMetaData()), serializer(rowIf->queryRowSerializer()), allocator(rowIf->queryRowAllocator()), deserializer(rowIf->queryRowDeserializer())
+    {
+        size32_t minSize = meta->getMinRecordSize();
+
+#ifdef STRESSTEST_SPILLING_ROWSTREAM
+        stressTest = activity.getOptBool("stressLookAhead", true); // JCSMORE temporarily defaulting to true
+        if (stressTest)
+        {
+            options.inMemMaxMem = minSize * 4;
+            options.writeAheadSize = options.inMemMaxMem * 2;
+            options.tempFileGranularity = options.inMemMaxMem * 4;
+            if (options.tempFileGranularity < 0x10000) // stop silly sizes (NB: this would only be set so small for testing!)
+                options.tempFileGranularity = 0x10000;
+        }
+#endif
+
+        if (minSize < 16)
+            minSize = 16;  // not too important, just using to cap inMemRows queue size
+        inMemRows.setCapacity(options.inMemMaxMem / minSize);
```

Comment:
This could get quite big.  Each entry is 24bytes, So the array could match inMemMaxMem.  probably not a problem...